OsmAnd/src/net/osmand/core/android/TileSourceProxyProvider.java

```diff
@@ -54,40 +63,46 @@ public boolean supportsNaturalObtainData() {
 	}
 
 	@Override
-	public SWIGTYPE_p_QByteArray obtainImageData(IMapTiledDataProvider.Request request) {
-		byte[] image = null;
-		try {
-			long requestTimestamp = System.currentTimeMillis();
-			int zoom = request.getZoom().swigValue();
-			int tileX = request.getTileId().getX();
-			int tileY = request.getTileId().getY();
-			String tileFilename = rm.calculateTileId(tileSource, tileX, tileY, zoom);
-			if (tileSource.couldBeDownloadedFromInternet()) {
-				final TileReadyCallback tileReadyCallback = new TileReadyCallback(tileSource, tileX, tileY, zoom);
-				rm.getMapTileDownloader().addDownloaderCallback(tileReadyCallback);
-				try {
-					while (tilesCache.getTileForMapSync(tileFilename, tileSource, tileX, tileY, zoom, true,
-							requestTimestamp) == null && System.currentTimeMillis() - requestTimestamp < IMAGE_LOAD_TIMEOUT) {
-						synchronized (tileReadyCallback.getSync()) {
-							if (tileReadyCallback.isReady()) {
-								break;
-							}
-							try {
-								tileReadyCallback.getSync().wait(50);
-							} catch (InterruptedException ignored) {
-							}
-						}
-					}
-				} finally {
-					rm.getMapTileDownloader().removeDownloaderCallback(tileReadyCallback);
-				}
-			} else {
-				tilesCache.get(tileFilename, requestTimestamp);
+	public SWIGTYPE_p_sk_spT_SkImage_const_t obtainImage(IMapTiledDataProvider.Request request) {
+		long requestTimestamp = System.currentTimeMillis();
+		int zoom = request.getZoom().swigValue();
+		int tileX = request.getTileId().getX();
+		int tileY = request.getTileId().getY();
+		float tileSize = tileSource.getTileSize();
+		double offsetY = 0;
+		if (tileSource.isEllipticYTile()) {
+			double latitude = MapUtils.getLatitudeFromTile(zoom, tileY);
+			double[] numberOffset = MapUtils.getTileEllipsoidNumberAndOffsetY(zoom, latitude, tileSource.getTileSize());
+			tileY = (int) numberOffset[0];
+			offsetY = numberOffset[1];
+		}
+		boolean shiftedTile = offsetY > 0;
+		byte[] first;
+		byte[] second;
+		first = getTileBytes(tileX, tileY, zoom, requestTimestamp);
+		if (first == null) {
+			return SwigUtilities.nullSkImage();
+		}
+		Bitmap bitmapResult = Bitmap.createBitmap((int)tileSize, (int)tileSize, Bitmap.Config.ARGB_8888);
```

Comment:
bitmapResult > resultTileBitmap
Do not `Bitmap.createBitmap` if `shiftedTile == false`